tokens/src/lib.rs

```diff
@@ -812,6 +890,170 @@ impl<T: Config> MultiReservableCurrency<T::AccountId> for Pallet<T> {
 	}
 }
 
+impl<T: Config> PalletFungibles::Inspect<T::AccountId> for Pallet<T> {
+	type AssetId = T::CurrencyId;
+	type Balance = T::Balance;
+
+	fn total_issuance(asset_id: Self::AssetId) -> Self::Balance {
+		Pallet::<T>::total_issuance(asset_id)
+	}
+	fn minimum_balance(asset_id: Self::AssetId) -> Self::Balance {
+		<Self as MultiCurrency<_>>::minimum_balance(asset_id)
+	}
+	fn balance(asset_id: Self::AssetId, who: &T::AccountId) -> Self::Balance {
+		Pallet::<T>::total_balance(asset_id, who)
+	}
+	fn reducible_balance(asset_id: Self::AssetId, who: &T::AccountId, keep_alive: bool) -> Self::Balance {
+		let a = Pallet::<T>::accounts(who, asset_id);
+		// Liquid balance is what is neither reserved nor locked/frozen.
+		let liquid = a.free.saturating_sub(a.frozen);
+		if frame_system::Pallet::<T>::can_dec_provider(who) && !keep_alive {
+			liquid
+		} else {
+			// `must_remain_to_exist` is the part of liquid balance which must remain to
+			// keep total over ED.
+			let must_remain_to_exist = T::ExistentialDeposits::get(&asset_id).saturating_sub(a.total() - liquid);
+			liquid.saturating_sub(must_remain_to_exist)
+		}
+	}
+	fn can_deposit(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DepositConsequence {
+		Pallet::<T>::deposit_consequence(who, asset_id, amount, &Pallet::<T>::accounts(who, asset_id))
+	}
+	fn can_withdraw(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> WithdrawConsequence<Self::Balance> {
+		Pallet::<T>::withdraw_consequence(who, asset_id, amount, &Pallet::<T>::accounts(who, asset_id))
+	}
+}
+
+impl<T: Config> PalletFungibles::Mutate<T::AccountId> for Pallet<T> {
+	fn mint_into(asset_id: Self::AssetId, who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
+		if amount.is_zero() {
+			return Ok(());
+		}
+		Pallet::<T>::try_mutate_account(who, asset_id, |account, _is_new| -> DispatchResult {
+			Pallet::<T>::deposit_consequence(who, asset_id, amount, &account).into_result()?;
+			account.free += amount;
+			Ok(())
+		})?;
+		<TotalIssuance<T>>::mutate(asset_id, |t| *t += amount);
+		Ok(())
+	}
+
+	fn burn_from(
+		asset_id: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> Result<Self::Balance, DispatchError> {
+		if amount.is_zero() {
+			return Ok(Self::Balance::zero());
+		}
+		let actual =
+			Pallet::<T>::try_mutate_account(who, asset_id, |account, _is_new| -> Result<T::Balance, DispatchError> {
+				let extra = Pallet::<T>::withdraw_consequence(who, asset_id, amount, &account).into_result()?;
+				let actual = amount + extra;
+				account.free -= actual;
+				Ok(actual)
+			})?;
+		<TotalIssuance<T>>::mutate(asset_id, |t| *t -= actual);
+		Ok(actual)
+	}
+}
+
+impl<T: Config> PalletFungibles::Transfer<T::AccountId> for Pallet<T> {
+	fn transfer(
+		asset_id: Self::AssetId,
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: T::Balance,
+		_keep_alive: bool,
```

Comment:
we shouldn't ignore keep_alive

Reply:
https://github.com/open-web3-stack/open-runtime-module-library/blob/7cd2c10d2e9ea44015fdc2ab2d029c7695dac91a/tokens/src/lib.rs#L519-L538
Tokens `transfer` not handle the `ExistenceRequirement`

Reply:
we need a version that handles it, and `transfer` can call it with `keep_alive: false`